Updates for FOTA

mike-scott opened the pull request:
There are a few cleanups, but the primary focus of these changes is to make various components of our FOTA system optional to allow for future incoming additions such as LWM2M for sensor data as well as device management.

mbolivar commented on src/main.c:
```diff
@@ -18,8 +18,10 @@
 
 /* Local helpers and functions */
 #include "tstamp_log.h"
+#if defined(CONFIG_BLUETOOTH)
```
What about the <bluetooth/conn.h> above?

mike-scott replied:
Er.. missed this comment.  What about the <bluetooth/conn.h> above?

mbolivar commented on src/main.c:
```diff
@@ -387,11 +375,15 @@ void main(void)
-	if (err) {
+#if defined(CONFIG_NET_TCP)
+	TC_PRINT("Initializing TCP\n");
+	if (tcp_init()) {
```
I like this change.

Couple of nits:

- I made start_tcp() because the TC_END_RESULT() macros print ``__func__``, so this made it more descriptive than printing a result from main(). Is it possible to keep that around?
- "preparation" typo in the commit log

mike-scott replied:
The main cleanup was that we were generating 2 "TC_END_REPORT(TC_FAIL);" events when start_tcp() has issues.  1 inside the start_tcp() and then again when we check err here.  Also, by the end of these patches we have a nice list of ***_init() functions each printing their own test case texts.  They're all in the main() function :/  should we put them all into their own functions?

mbolivar replied:
Doesn't seem too important to do that; maybe a cleanup for another day.

mbolivar commented on Kconfig:
```diff
@@ -69,8 +69,15 @@ config SYS_LOG_FOTA_LEVEL
 	  - 4 DEBUG: write SYS_LOG_DBG in addition to previous levels
 
+config FOTA_BLUEMIX
+	bool "Enable sensor transmittal via MQTT/IBM Bluemix"
```
maybe s/sensor transmittal via/sending temperature sensor data to/?

Or, in preparation for the later patch, maybe "exchanging data with IBM Bluemix via MQTT"?

mbolivar commented on Kconfig:
```diff
@@ -69,8 +69,15 @@ config SYS_LOG_FOTA_LEVEL
+	bool "Enable sensor transmittal via MQTT/IBM Bluemix"
+	default n
+	help
+	  This option adds logic for sending MQTT/Bluemix data
```
Eventually we'll send and receive, so maybe "exchanging" instead of "sending"?

mbolivar commented on src/lib/bluemix.c:
```diff
@@ -43,6 +48,17 @@
 		 CONFIG_FOTA_BLUEMIX_DEVICE_TYPE, ctx->bm_id)
 
+#define BLUEMIX_STACK_SIZE 1024
+char bluemix_thread_stack[BLUEMIX_STACK_SIZE];
```
Maybe make this static while you're here?

mbolivar commented on src/lib/bluemix.h:
```diff
@@ -78,7 +78,8 @@ struct bluemix_ctx {
 
-int bluemix_init(struct bluemix_ctx *ctx);
+int bluemix_init(void);
+int bluemix_start(struct bluemix_ctx *ctx);
```
This is a little strange. It seems like bluemix_start() and bluemix_fini() still get called from the service thread, but the application never touches them anymore.

Since this is being reorganized to have bluemix_init() start a thread, perhaps the application just needs a hook to ask the thread to stop and then join on it? It's not clear bluemix_start() / bluemix_fini() should still be public API anymore.

mbolivar commented on src/lib/bluemix.c:
```diff
@@ -322,3 +338,170 @@ int bluemix_pub_status_json(struct bluemix_ctx *ctx,
+		 * an MCU temperature sensor or encounter errors
+		 * reading it, use these values as defaults.
+		 */
+		ret = get_temp_sensor_data(mcu_temp_sensor_dev,
```
It's kind of messy to move this into the bluemix core.

What about passing a callback to bluemix_init() that gets invoked whenever the service thread wakes up and successfully connects? Then this code can move out of here and into the application callback.

The application itself could have a source file that's obj-$(CONFIG_FOTA_BLUEMIX), which calls bluemix_init() with the callback at CONFIG_APPLICATION_INIT_PRIORITY time.

mike-scott replied:
Yeah, I was kind of noodling on how best to abstract a sensor interface.   But we couldn't leave it in the main.c while moving Bluemix stuff out. :/   I think this will all go away eventually and we'll be using LWM2M. :P

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -959,3 +968,65 @@ int hawkbit_ddi_poll(void)
+		}
+#endif
+
+#if defined(CONFIG_NET_TCP)
```
Is hawkbit ever selected without CONFIG_NET_TCP?

mike-scott replied:
Hrm, good point.. perhaps I can remove this here, and instead add "select NET_TCP" in our Kconfig option for FOTA_DM_BACKEND_HAWKBIT?

mbolivar replied:
What about adding "depends on CONFIG_NET_TCP" to FOTA_DM_BACKEND_HAWKBIT?

mike-scott replied:
Let's go with that.

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -241,7 +250,7 @@ static int hawkbit_device_acid_update(hawkbit_dev_acid_t type,
 }
 
-int hawkbit_init(void)
+int hawkbit_start(void)
```
Seems like this should be static

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -959,3 +968,65 @@ int hawkbit_ddi_poll(void)
+		}
+		SYS_LOG_ERR("Unsupported device management backend");
+
+#if defined(CONFIG_NET_TCP)
```
Same question here about CONFIG_NET_TCP

mbolivar commented on src/lib/tcp.c:
```diff
@@ -159,22 +157,12 @@ static void tcp_received_cb(struct net_context *context,
 		}
+
+		pos = net_pkt_appdata(pkt) - net_pkt_ip_data(pkt);
+		if (!net_frag_read(pkt->frags, pos, &pos, net_pkt_appdatalen(pkt), ctx->read_buf + ctx->read_bytes))
```
80 columns

mbolivar commented on src/lib/tcp.c:
```diff
@@ -197,20 +197,18 @@ int tcp_init(void)
 	/* Add delays so DHCP can assign IP */
 	/* TODO: add a timeout/retry */
-	SYS_LOG_INF("Waiting for DHCP ");
+	SYS_LOG_INF("Waiting for DHCP ...");
```
Maybe remove the trailing spaces here and below, too?

mike-scott commented:
Most items were taken into account with V2:
https://hastebin.com/puqeroyuxa.cs

outstanding questions:
- cleanup handling of the temp sensor code (you got your temp sensor code in my bluemix!)
- do we need to move all of the ***_init() functions out of main() into their own functions for log failure clarity.

mike-scott commented:
+ another outstanding question:
- Should I remove the NET_TCP checks in hawkbit.c in favor of having the Kconfig option select NET_TCP.

mike-scott commented:
Ugh, I noticed that I updated the k_thread_create API when I moved the stack init code without specifically calling that out.  I may create a separate patch for that change after moving the stack init code.

mike-scott commented:
Fixing checkpatch and a compile issue on BT devices

mike-scott commented:
V3 diff: https://hastebin.com/kineyozapi.cs

- Fixed checkpatch errors
- Fixed compile problem on bluetooth devices

mbolivar commented:
It's not guarded by a Bluetooth ifdef

On May 18, 2017 8:09 PM, "Michael Scott" <notifications@github.com> wrote:

> *@mike-scott* commented on this pull request.
> ------------------------------
>
> In src/main.c
> <https://github.com/Linaro/zephyr-fota-hawkbit/pull/83#discussion_r117381162>
> :
>
> > @@ -18,8 +18,10 @@
>
>  /* Local helpers and functions */
>  #include "tstamp_log.h"
> +#if defined(CONFIG_BLUETOOTH)
>
> Er.. missed this comment. What about the <bluetooth/conn.h> above?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Linaro/zephyr-fota-hawkbit/pull/83#discussion_r117381162>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACJEAOE15AsWb4vRCScTEaGLYj9vPjfks5r7N2-gaJpZM4Nf0Ac>
> .
>

mike-scott commented:
V4 diff: https://hastebin.com/lebafaceho.cpp

- Removed the k_thread_create update that slipped in.  There's no change to the stack creation except for static var additions.
- Made the main.c TCP_NET guard section match the other since there was a checkpatch complaint

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -959,3 +968,65 @@ int hawkbit_ddi_poll(void)
+			/* restart the failed attempt counter */
+			hawkbit_failures = 0;
+		}
+		SYS_LOG_ERR("Unsupported device management backend");
```
This should be deleted

doanac commented:
Feel free to ignore CI stuff. I just noticed the webhooks weren't working on this project properly. I'm trying to sort it out.

mike-scott commented:
V5 diff: https://hastebin.com/jedifubaqa.cs

Changes:
- added last 2 patches to clean up "err" compile warning and the API change for k_thread_create

mike-scott commented:
(Also pulled in the changes @mbolivar made)

mbolivar commented on src/main.c:
```diff
@@ -38,25 +45,17 @@
 static bool bt_connection_state = false;
 
+/* external method for reading bt state */
+bool bt_connection_state_get(void)
```
There shouldn't be any dependencies in lib/ to FOTA code outside of lib/. Can this be moved?

mike-scott replied:
@mbolivar I'm going to add another patch to this series where all of the CONFIG_BLUETOOTH logic is moved out of main.c as well.  The new file will be src/lib/bluetooth.c and use the header file I created before (bluetooth.h).  That way Bluemix and Hawkbit can rely in on that.

mike-scott replied:
NIX that.  Each of these files in src/lib needs to work totally independently.  I'll look at adding a patch which changes how both hawkbit.c and bluemix.c know that the bluetooth interface is up/enabled via the net management events API.  We may need to upstream a patch or 2 to support this tho.

mike-scott commented:
V6 diff: https://pastebin.com/raw/ZuRFGe4c

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -1022,5 +1031,17 @@ int hawkbit_init(void)
+#if defined(CONFIG_NET_MGMT_EVENT)
+	/* Subscribe to NET_IF_UP if interface is not ready */
+	if (!atomic_test_bit(iface->flags, NET_IF_UP)) {
+		net_mgmt_init_event_callback(&cb, event_iface_up,
```
I really like using this API. But it seems like this patch makes an improvement by assuming that there is exactly one network interface on the board. It's an improvement so I think it should be merged, but how can we make this better in the future?

How about having hawkbit_init take a `struct net_if*` argument, stashing it in a context struct, and comparing it to the net_if we get in event_iface_up?

That would also help fix this code in case net_if_get_default() returns NULL (unlikely in practice, but hey).

mike-scott replied:
Yes, this is a good point.  Practically every Zephyr sample does it this way, but in reality we could have multiple interfaces as you point out.  Probably best to add the parameter as you mention and let the caller to decide which interface to use.

mbolivar replied:
It also seems like we could add callbacks on NET_EVENT_IF_DOWN in future patches, to set connection_ready back to false.

mbolivar commented on src/main.c:
```diff
@@ -385,33 +149,32 @@ void main(void)
-	k_thread_spawn(&bluemix_thread_stack[0], BLUEMIX_STACK_SIZE,
-			(k_thread_entry_t) bluemix_service,
-			NULL, NULL, NULL, K_PRIO_COOP(7), 0, K_NO_WAIT);
+	TC_END_REPORT(TC_PASS);
```
This should be removed. The real successful end of report should come from blink_led().

mbolivar commented on src/lib/hawkbit.c:
```diff
@@ -959,3 +971,78 @@ int hawkbit_ddi_poll(void)
+#if defined(CONFIG_NET_MGMT_EVENT)
+	/* Subscribe to NET_IF_UP if interface is not ready */
+	if (!atomic_test_bit(iface->flags, NET_IF_UP)) {
+		net_mgmt_init_event_callback(&cb, event_iface_up,
```
I tested this and got the following error on Nitrogen:

```
***** BOOTING ZEPHYR OS v1.7.99 - BUILD: May 24 2017 13:51:11 *****
[0000000] [fota/main] [INF] main: Linaro FOTA example application
[0000000] [fota/main] [INF] main: Device: 96b_nitrogen, Serial: 5b0f7d64
tc_start() - Running Built in Self Test (BIST)
Setting Bluetooth MAC
[0000010] [fota/bt_storage] [DBG] bt_storage_init: Bluetooth storage driver registered
===================================================================
PASS - bt_storage_init.
Enabling Bluetooth
[0000030] [bt] [INF] show_dev_info: Identity: d6:e7:5b:0f:7d:64 (random)
[0000040] [bt] [INF] show_dev_info: HCI: version 5.0 (0x09) revision 0x0000, manufacturer 0xffff
[0000050] [bt] [INF] show_dev_info: LMP: version 5.0 (0x09) subver 0xffff
[0000050] [bt] [WRN] bt_pub_key_gen: ECC HCI commands not available
===================================================================
PASS - main.
Registering Bluetooth LE connection callbacks
===================================================================
PASS - ipss_init.
Advertising Bluetooth IP Profile
===================================================================
PASS - ipss_advertise.
Initializing TCP
===================================================================
PASS - main.
Initializing Hawkbit backend
[0000100] [fota/hawkbit] [INF] hawkbit_start: ACID: current -1, update 4
[0000110] [fota/hawkbit] [INF] hawkbit_start: Current boot status ff
[0000110] [fota/hawkbit] [INF] hawkbit_start: Updated boot status to 1
[0000230] [fota/hawkbit] [DBG] hawkbit_start: Erased flash bank at offset 3c000
[0000240] [fota/hawkbit] [INF] hawkbit_start: ACID updated, current 4, update 4
[0000250] [fota/hawkbit] [INF] hawkbit_service: Starting FOTA Service Thread
===================================================================
PASS - main.
Initializing Bluemix Client service
[0000260] [fota/bluemix] [INF] temp_init: Found MCU temperature sensor fota-mcu-temp
[0000270] [fota/bluemix] [INF] temp_init: Did not find off-chip temperature sensor fota-offchip-temp
===================================================================
PROJECT EXECUTION SUCCESSFUL
Blinking LED
[0000740] [fota/main] [INF] connected: BT LE Connected
===================================================================
PASS - blink_led.
===================================================================
PROJECT EXECUTION SUCCESSFUL
[0003290] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0011310] [fota/tcp] [ERR] tcp_connect_context: Cannot connect to server: fe80::d4e7:0:0:1:1883 (-60)
[0011310] [fota/bluemix] [DBG] bluemix_service: Failed Bluemix init - attempt 1
```
That was weird; I would have expected this to work now that the callbacks are in place.

Here's what I saw on Carbon:

```
[0003529] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0006537] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0009545] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0012553] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0015561] [fota/bluemix] [DBG] bluemix_service: Network interface is not ready
[0016148] [fota/main] [INF] connected: BT LE Connected
[0019406] [fota/bluemix] [DBG] bluemix_service: Read MCU temp sensor: 19.0C
[0019414] [fota/bluemix] [DBG] publish_message: topic:iot-2/type/96b_carbon/id/96b_carbon-605f82f9/evt/status/fmt/json
[0019425] [fota/bluemix] [DBG] publish_message: message:{"d":{"mcutemp":19}}
```

On K64F, I use DHCP, so waiting for IP assignment before any of the bluemix code runs means it worked there, too.

mbolivar commented:
Noticed a couple of issues during testing.
